docs/proposals/20231227-koordlet-resctrl-qos-enhance.md

```diff
@@ -0,0 +1,468 @@
+    closid := GetClosId(annotation, label)
+    app, err := p.engine.GetApp(podid)
+    updateContainerContext(podid, containerid, app.closid, app.Hooks)
+}
+
+// UpdatePodTaskIds will update new taskids to resctrl file system
+func (p *plugin) UpdatePodTaskIds(proto protocol.HooksProtocol) error {
+    // 1. retrieve task ids for each slo and each specific LLC/MBA request pod by consume pod taskids
+    // 2. add the new related task ids in resctrl groups
+}
+```
+
+**PodContext**
+```go
+
+func (p *PodContext) NriDone(executor resourceexecutor.ResourceUpdateExecutor) {
+    if p.executor == nil {
+        p.executor = executor
+    }
+    p.injectForExt()
+    p.Update()
+}
+
+func (p *PodContext) NRIRemoveDone() {
+    if p.executor == nil {
+        p.executor = executor
+    }
+    p.removeForExt()
+    p.Update()
+}
+
+// removeForExt will handle NRI remove/clean operation
+func (p *PodContext) removeForExt() {
+    
+}
+func injectResctrl(closid string, schemata string, a *audit.EventHelper, e resourceexecutor.ResourceUpdateExecutor) (resourceexecutor.ResourceUpdater, error) {
+    // for specific pods, create control group and monitor group and update schemata
+}
+
+func removeResctrl(closid string, a *audit.EventHelper, e resourceexecutor.ResourceUpdateExecutor) (resourceexecutor.ResourceUpdater, error) {
+    // for specific pods, create control group and monitor group and update schemata
+}
+```
+
+**ContainerContext**
+```go
+func (c *ContainerContext) NriDone(executor resourceexecutor.ResourceUpdateExecutor) (*api.ContainerAdjustment, *api.ContainerUpdate, error) {
+    ......
+    if c.Response.Resources.Resctrl != nil {
+        // adjust OCI spec
+        adjust.SetLinuxRDTClass(c.Response.Resources.Resctrl.Closid)
+        adjust.Hooks = c.Response.Resources.Resctrl.Hooks
+    }
+}
+```
+
+**Resources**
+
+```go
+type Resctrl struct {
+    Schemata string
+    Hooks string
+    Closid string
+}
+
+type Resources struct {
+    // origin resources
+    CPUShares   *int64
+    CFSQuota    *int64
+    CPUSet      *string
+    MemoryLimit *int64
+
+    // extended resources
+    CPUBvt *int64
+    Resctrl *Resctrl
+}
+```
+
+**ResctrlUpdater**
+
+We will enhance current ResctrlUpdate to support directly overwrite schemata. Current ResctrlUpdater doesn't support override schemata with LLC and MBA together directly and doesn't consider NUMA when apply schemata .
+
+```go
+func NewResctrlSchemata(group, schemata string) ResourceUpdater {
+
+}
+
+func NewResctrlSchemataWithNuma(group, resctrl Resctrl) ResourceUpdater {
+
+}
+
+```
+
+**podsInformer**
+
+podsInformer will add a new interface to get all pods' taskids, Resctl runtime hook Reconciler will register callback to consume these information and write new taskIds to resctrl contrl group and monitor group.
+
+```go
+
+type PodMeta struct {
+    Pod       *corev1.Pod
+    CgroupDir string
+    // Add new field, todo: memory usage
+    ContainerTaskIds map[string][]uint64
```

Comment:
check whether save task id by default(no matter the plugin enable or not), especially the performance under cgroupV1

Reply:
I didn't find any clues that reading cgroupfs pseudo-file has performance impact. However, the impact of reading the tasks file frequently depends on the scale of the system and the frequency of the reads. If you are on a system with a large number of tasks or if you are performing the reads at a very high rate, it might contribute to a small amount of overhead. We'd better control the read rate. What about add an rate limiter when we implement this feature ?